optuna/visualization.py

```diff
@@ -92,6 +99,164 @@ def _get_intermediate_plot(study):
+            for y_i, y_param in enumerate(sorted_params):
+                if x_param == y_param:
+                    figure.add_trace(go.Scatter(), row=y_i + 1, col=x_i + 1)
+                else:
+                    sub_plots = _generate_contour_subplot(
+                        trials, x_param, y_param, study.direction)
+                    contour = sub_plots[0]
+                    scatter = sub_plots[1]
+                    contour.update(showscale=showscale)  # showscale's default is True
+                    if showscale:
+                        showscale = False
+                    figure.add_trace(contour, row=y_i + 1, col=x_i + 1)
+                    figure.add_trace(scatter, row=y_i + 1, col=x_i + 1)
+                figure.update_xaxes(range=param_values_range[x_param],
+                                    row=y_i + 1, col=x_i + 1)
+                figure.update_yaxes(range=param_values_range[y_param],
+                                    row=y_i + 1, col=x_i + 1)
+                if x_i == 0:
+                    figure.update_yaxes(title_text=y_param, row=y_i + 1, col=x_i + 1)
+                if y_i == len(sorted_params) - 1:
+                    figure.update_xaxes(title_text=x_param, row=y_i + 1, col=x_i + 1)
+
+    return figure
+
+
+def _generate_contour_subplot(trials, x_param, y_param, direction):
+    # type: (List[FrozenTrial], str, str, StudyDirection) -> Tuple[Contour, Scatter]
+    x_indexes = sorted(list({t.params[x_param] for t in trials if x_param in t.params}))
+    y_indexes = sorted(list({t.params[y_param] for t in trials if y_param in t.params}))
+    if len(x_indexes) < 2:
+        logger.warning('Param {} unique value length is less than 2.'.format(x_param))
+        return go.Contour(), go.Scatter()
+    if len(y_indexes) < 2:
+        logger.warning('Param {} unique value length is less than 2.'.format(y_param))
+        return go.Contour(), go.Scatter()
+    z = [[float('nan') for _ in range(len(x_indexes))] for _ in range(len(y_indexes))]
+
+    x_values = []
+    y_values = []
+    for trial in trials:
+        if x_param not in trial.params or y_param not in trial.params:
+            continue
+        x_values.append(trial.params[x_param])
+        y_values.append(trial.params[y_param])
+        x_i = x_indexes.index(trial.params[x_param])
+        y_i = y_indexes.index(trial.params[y_param])
+        if isinstance(trial.value, int):
+            value = float(trial.value)
+        elif isinstance(trial.value, float):
+            value = trial.value
+        else:
+            raise ValueError(
+                'Trial{} has COMPLETE state, but its value is non-numeric.'.format(trial.number))
+        z[y_i][x_i] = value
+
+    contour = go.Contour(
+        x=x_indexes, y=y_indexes, z=z,
+        colorbar={'title': 'Objective Value'},
+        colorscale='blues',
+        connectgaps=True,
+        contours_coloring='heatmap',
+        hoverinfo='none',
+        line_smoothing=1.3,
```

Comment:
I'm unfamiliar with `plotly`, so I'm not sure whether this default value is good for a wide range of use-cases.
What do you think?

Reply:
From the reference below, the range of line_smoothing is 0.0-1.3 and default is 1.0. I chose 1.3 because I thought it would be better to have a smoother line.

https://plot.ly/python/reference/#contour-line-smoothing

Reply:
I see. Thank you for your explanation.